README
NAME
    Catalyst::View::Component::SubInclude - Use subincludes in your Catalyst
    views

VERSION
    Version 0.11

SYNOPSIS
      package MyApp::View::TT;
      use Moose;

      extends 'Catalyst::View::TT';
      with 'Catalyst::View::Component::SubInclude';

      __PACKAGE__->config( subinclude_plugin => 'SubRequest' );

    Then, somewhere in your templates:

      [% subinclude('/my/widget') %]
      [% subinclude_using('SubRequest', '/page/footer') %]

DESCRIPTION
    "Catalyst::View::Component::SubInclude" allows you to include content in
    your templates (or, more generally, somewhere in your view's "render"
    processing) which comes from another action in your application. It's
    implemented as a Moose::Role, so using Moose in your view is required.

    Simply put, it's a way to include the output of a Catalyst sub-request
    somewhere in your page.

    It's built in an extensible way so that you're free to use sub-requests,
    Varnish ESI (<http://www.catalystframework.org/calendar/2008/17>) or any
    other sub-include plugin you might want to implement.

STASH FUNCTIONS
    This component does its magic by exporting a "subinclude" coderef entry
    to the stash. This way, it's easily accessible by the templates (which
    is the most common use-case).

  "subinclude( $path, @args )"
    This will render and return the body of the included resource (as
    specified by $path) using the default subinclude plugin.

  "subinclude_using( $plugin, $path, @args )"
    This will render and return the body of the included resource (as
    specified by $path) using the specified subinclude plugin.

    The "subinclude" function above is implemented basically as a shortcut
    which calls this function using the default plugin as the first
    parameter.

SUBINCLUDE PLUGINS
    The module comes with two subinclude plugins: SubRequest, Visit and ESI.

    By default, the "SubRequest" plugin will be used. This can be changed in
    the view's configuration options (either in the config file or in the
    view module itself).

        __PACKAGE__->config(
            subinclude_plugin => 'ESI',
            subinclude => {
                'SubRequest' => {
                    keep_stash => 1,
                },
                'HTTP::POST' => {
                    class => 'HTTP',
                    http_method => 'POST',
                    ua_timeout => '10',
                    uri_map => {
                        '/foo/' => 'http://www.foo.com/',
                    },
                },
            },
        );

    You can change each plugins' configuration through the keys in the
    'subinclude' config key (example above)

  "set_subinclude_plugin( $plugin )"
    This method changes the current active subinclude plugin in runtime. It
    expects the plugin suffix (e.g. "ESI" or "SubRequest") or a
    fully-qualified class name in the
    "Catalyst::View::Component::SubInclude" namespace.

  Writing plugins
    If writing your own plugin, keep in kind plugins are required to
    implement a class method "generate_subinclude" with the following
    signature:

      sub generate_subinclude {
          my ($class, $c, @args) = @_;
      }

    The default plugin is stored in the "subinclude_plugin" which can be
    changed in runtime. It expects a fully qualified class name.

SEE ALSO
    Catalyst::Plugin::SubRequest, Moose::Role, Moose,
    <http://www.catalystframework.org/calendar/2008/17>

BUGS
    Please report any bugs or feature requests to
    "bug-catalyst-view-component-subinclude at rt.cpan.org", or through the
    web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-View-Component-
    SubInclude>. I will be notified, and then you'll automatically be
    notified of progress on your bug as I make changes.

AUTHOR
    Nilson Santos Figueiredo Junior, "<nilsonsfj at cpan.org>"

CONTRIBUTORS
    Tomas Doran (t0m) "<bobtfish@bobtfish.net".

    Vladimir Timofeev, "<vovkasm at gmail.com>".

    Wallace Reis (wreis) "<wreis@cpan.org>".

SPONSORSHIP
    Development sponsored by Ionzero LLC <http://www.ionzero.com/>.

COPYRIGHT & LICENSE
    Copyright (C) 2010 Nilson Santos Figueiredo Junior and the above
    contributors.

    Copyright (C) 2009 Nilson Santos Figueiredo Junior.

    Copyright (C) 2009 Ionzero LLC.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Catalyst-View-Component-SubInclude

0.11  Tue 31 May 20:11:00 2010
      - Document and test the keep-stash option (Jose Luis Martinez)

0.10  Tue 14 Dec 14:32:00 2010
      - Add HTTP support (Wallace Reis)

0.09  Thu 10 June 21:24:00 2010
      - Add SSI support (Vladimir Timofeev)
      - Additional documentation.

0.08  Sun 16 May 15:52:00 2010
      - Released 0.07_03 with no further changes.

0.07_03 Fri Feb 19 00:31:00 2010
      - Autogenerate README file from Pod.
      - Fix missing test_requires dependency on parent by dropping use of
         that module.

0.07_02 Fri Feb 19 00:11:00 2010
        - Fix missing test_requires dependency on Catalyst::Action::RenderView.

0.07_01 Mon Feb 15 22:41:54 2010
        - It is now possible to configure plugins, and the subrequest
          and visit plugins can be configured to not replace the stash
          before redispatching.
        - Fix Visit plugin to work with newer Catalyst versions.
        - Use uri_for_action instead of calling methods in the dispatcher
          directly when possible.
        - Call pulbic rather than private methods in the dispatcher in the
          subrequest backend.
        - Plugins which are used now generate an instance for each plugin
          which is subsequently cached.
        - Fix useage of namespace::clean
        - Proper test application bundled with the distribution.
        - Proper tests for all engines in the distribution.
        - Use Catalyst::Component::ContextClosure instead of weakening
          the context manually.
        - Make pod tests only run in author mode.
        - General code cleanup.

0.07    Mon Jul 27 20:45:42 2009
        - Fix some issues related to Args/CaptureArgs handling in the SubRequest
          and Visit plugins.

0.06
        - Weaken $c to avoid a memory leak.

0.05    Tue Apr 21 19:16:00 2009
        - Change attribute constraint from ClassName to Str to fix issue with
          newer Catalyst versions

0.04    Wed Feb 11 00:38:02 2009
        - Recreate dist because the CPAN indexer doesn't like tarballs generated
          on Windows.

0.03    Fri Feb  6 17:58:50 2009
        - Add "subinclude_using"
        - Tidy up dist (added README and this Changes file)

0.02    Thu Feb  5 23:26:54 2008
        - Fixed several issues related to Chained actions
        - Added the Visit subinclude plugin

0.01    Fri Jan 30 21:00:00 2009
        First public version.



t/lib/KeepStash.pm
package KeepStash;

use strict;
use warnings;

use Catalyst::Runtime '5.70';

use base qw/Catalyst/;
use Catalyst qw/
    SubRequest
/;

our $VERSION = '0.01';


__PACKAGE__->config( name => 'KeepStash' );

__PACKAGE__->setup();

1;


t/lib/KeepStash/View/TT.pm
package KeepStash::View::TT;
use Moose;

extends 'Catalyst::View::TT';
with 'Catalyst::View::Component::SubInclude';

__PACKAGE__->config(
    TEMPLATE_EXTENSION => '.tt',
    subinclude_plugin => 'SubRequest',
    subinclude_plugin => 'Visit',
    subinclude => {
        'SubRequest' => {
            keep_stash => 1
        },
        'Visit' => {
            keep_stash => 1
        }
    },
);

1;


t/lib/ESITest/Controller/Root.pm
package ESITest::Controller::Root;

use strict;
use warnings;
use base 'Catalyst::Controller';

__PACKAGE__->config->{namespace} = '';

sub auto : Private {
    my ( $self, $c ) = @_;
    $c->stash->{'current_view'} = 'TT';
    return 1;
}

sub index :Path Args(0) {}

sub base : Chained('/') PathPart('') CaptureArgs(0) {}

sub time_include : Chained('base') PathPart('time') Args(0) {
    my ( $self, $c ) = @_;
    my $params = $c->req->params;

    $c->stash->{current_time} = localtime();
    
    my $additional = '';
    for my $key (keys %$params) {
        $additional .= "| $key = $params->{$key} | "
    }

    $c->stash->{additional} = $additional;
    
}

sub capture : Chained('base') PathPart('') CaptureArgs(1) {
    my ( $self, $c, $arg ) = @_;
    $c->log->debug("Capture: $arg") if $c->debug;
    $c->stash->{additional} = "Capture Arg: $arg";
}

sub time_args : Chained('capture') PathPart('time') Args(0) {
    my ( $self, $c ) = @_;
    my $params = $c->req->params;

    $c->stash->{current_time} = localtime();

    my $additional = $c->stash->{additional};
    for my $key (keys %$params) {
        $additional .= "| $key = $params->{$key} | "
    }

    $c->stash->{additional} = $additional;

    $c->stash->{template} = 'time_include.tt';
}

sub time_args_with_args : Chained('capture') PathPart('time') Args(1) {
    my ( $self, $c, $arg ) = @_;
    my $params = $c->req->params;

    $c->stash->{current_time} = localtime();

    my $additional = $c->stash->{additional};
    for my $key (keys %$params) {
        $additional .= " | $key = $params->{$key} | "
    }

    $additional .= " Action Arg: $arg ";

    $c->stash->{additional} = $additional;

    $c->stash->{template} = 'time_include.tt';
}

sub time_args_without_capture : Chained('base') PathPart('time') Args(1) {
    my ( $self, $c, $arg ) = @_;
    my $params = $c->req->params;

    $c->stash->{current_time} = localtime();

    my $additional = '';
    for my $key (keys %$params) {
        $additional .= " | $key = $params->{$key} | "
    }

    $additional .= " Action Arg: $arg ";

    $c->stash->{additional} = $additional;

    $c->stash->{template} = 'time_include.tt';
}

sub time_args_no_chained : Path('time_args_no_chained') Args {
    my ($self, $c, @args) = @_;

    my $params = $c->req->params;

    $c->stash->{current_time} = localtime();

    my $additional = '';
    for my $key (keys %$params) {
        $additional .= " | $key = $params->{$key} | "
    }

    $additional .= " No Chained Args: " . join ', ', @args;

    $c->stash->{additional} = $additional;

    $c->stash->{template} = 'time_include.tt';
}

sub http : Chained('base') PathPart('') CaptureArgs(0) {
    pop->stash->{'current_view'} = 'TTWithHTTP';
}

sub http_cpan : Chained('http') Args(0) {}

sub http_github : Chained('http') Args(0) {}

sub end : ActionClass('RenderView') {}

1;


lib/Catalyst/View/Component/SubInclude/SSI.pm
package Catalyst::View::Component::SubInclude::SSI;
use Moose;
use namespace::clean -except => 'meta';

=head1 NAME

Catalyst::View::Component::SubInclude::SSI - Server Side Includes (SSI) plugin for C::V::Component::SubInclude

=head1 VERSION

Version 0.09

=cut

our $VERSION = '0.09';
$VERSION = eval $VERSION;

=head1 SYNOPSIS

In your view class:

  package MyApp::View::TT;
  use Moose;

  extends 'Catalyst::View::TT';
  with 'Catalyst::View::Component::SubInclude';

  __PACKAGE__->config( subinclude_plugin => 'SSI' );

Then, somewhere in your templates:

  [% subinclude('/my/widget') %]

=head1 DESCRIPTION

C<Catalyst::View::Component::SubInclude::SSI> renders C<subinclude> calls as 
Server Side Includes (SSI) include directives. This is a feature implemented by 
Apache (L<http://httpd.apache.org/>), nginx (L<http://wiki.nginx.org/Main>)
and many other web servers which allows cache-efficient uses of includes.

=head1 METHODS

=head2 C<generate_subinclude( $c, $path, @args )>

Note that C<$path> should be the private action path - translation to the public
path is handled internally. After translation, this will roughly translate to 
the following code:

  my $url = $c->uri_for( $translated_path, @args )->path_query;
  return '<!--#include virtual="$url" -->';

Notice that the stash will always be empty. This behavior could be configurable
in the future through an additional switch - for now, this behavior guarantees a
common interface for plugins.

=cut

sub generate_subinclude {
    my ($self, $c, $path, @params) = @_;

    my $uri = $c->uri_for_action( $path, @params );

    return '<!--#include virtual="' . $uri->path_query . '" -->';
}

=head1 SEE ALSO

L<Catalyst::View::Component::SubInclude|Catalyst::View::Component::SubInclude>, 

=head1 AUTHOR

Vladimir Timofeev, C<< <vovkasm at gmail.com> >>

=head1 COPYRIGHT & LICENSE

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

__PACKAGE__->meta->make_immutable;
1;


lib/Catalyst/View/Component/SubInclude/HTTP.pm
package Catalyst::View::Component::SubInclude::HTTP;

use Moose;
use namespace::clean -except => 'meta';
use LWP::UserAgent;
use List::MoreUtils 'firstval';
use URI;

our $VERSION = '0.01';
$VERSION = eval $VERSION;

has http_method => (
    isa => 'Str', is => 'ro', default => 'GET',
);

has ua_timeout => (
    isa => 'Int', is => 'ro', default => 10,
);

has base_uri => (
    isa => 'Str', is => 'ro', required => 0,
);

has uri_map => (
    isa => 'HashRef', is => 'ro', required => 0,
);

has user_agent => (
    is => 'ro', lazy => 1, builder => '_build_user_agent',
);

sub _build_user_agent {
    my $self = shift;
    return LWP::UserAgent->new(
        agent => ref($self),
        timeout => $self->ua_timeout,
    );
}

sub generate_subinclude {
    my ($self, $c, $path, $args) = @_;
    my $error_msg_prefix = "SubInclude for $path failed: ";
    my $base_uri = $self->base_uri || $c->req->base;
    my $uri_map = $self->uri_map || { q{/} => $base_uri };
    $base_uri = $uri_map->{ firstval { $path =~ s/^$_// } keys %$uri_map };
    $base_uri =~ s{/$}{};
    my $uri = URI->new(join(q{/}, $base_uri, $path));
    my $req_method = q{_} . lc $self->http_method . '_request';

    my $response;
    if ( $self->can($req_method) ) {
        $response = $self->$req_method($uri, $args);
    }
    else {
        confess $self->http_method . ' not supported';
    }
    if ($response->is_success) {
        return $response->content;
    }
    else {
        $c->log->info($error_msg_prefix . $response->status_line);
        return undef;
    }
}

sub _get_request {
    my ( $self, $uri, $args) = @_;
    $uri->query_form($args);
    return $self->user_agent->get($uri);
}

sub _post_request {
    my ( $self, $uri, $args ) = @_;
    return $self->user_agent->post($uri, $args);
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

Catalyst::View::Component::SubInclude::HTTP - HTTP plugin for C::V::Component::SubInclude

=head1 SYNOPSIS

In your view class:

    package MyApp::View::TT;
    use Moose;

    extends 'Catalyst::View::TT';
    with 'Catalyst::View::Component::SubInclude';

    __PACKAGE__->config(
        subinclude_plugin => 'HTTP::GET',
        subinclude => {
            'HTTP::GET' => {
                class => 'HTTP',
                http_method => 'GET',
                ua_timeout => '10',
                uri_map => {
                    '/my/' => 'http://localhost:5000/',
                },
            },
            'HTTP::POST' => {
                class => 'HTTP',
                http_method => 'POST',
                ua_timeout => '10',
                uri_map => {
                    '/foo/' => 'http://www.foo.com/',
                },
            },
        },
    );

Then, somewhere in your templates:

    [% subinclude('/my/widget') %]
    ...
    [% subinclude_using('HTTP::POST', '/foo/path', { foo => 1 }) %]

=head1 DESCRIPTION

C<Catalyst::View::Component::SubInclude::HTTP> does HTTP requests (currently
using L<LWP::UserAgent>) and uses the responses to render subinclude contents.

=head1 CONFIGURATION

The configuration is passed in the C<subinclude> key based on your plugin name
which can be arbitrary.

=over

=item class

Required just in case your plugin name differs from C<HTTP>.

=item http_method

Accepts C<GET> and C<POST> as values. The default one is C<GET>.

=item user_agent

This lazily builds a L<LWP::UserAgent> obj, however you can pass a different
user agent obj that implements the required API.

=item ua_timeout

User Agent's timeout config param. Defaults to 10 seconds.

=item uri_map

This expects a HashRef in order to map paths to different URLs.

=item base_uri

Used only if C<uri_map> is C<undef> and defaults to C<< $c->request->base >>.

=back

=head1 METHODS

=head2 C<generate_subinclude( $c, $path, $args )>

Note that C<$path> should be the relative path.

=head1 SEE ALSO

L<Catalyst::View::Component::SubInclude|Catalyst::View::Component::SubInclude>

=head1 AUTHOR

Wallace Reis C<< <wreis@cpan.org> >>

=head1 SPONSORSHIP

Development sponsored by Ionzero LLC L<http://www.ionzero.com/>.

=head1 COPYRIGHT & LICENSE

Copyright (c) 2010 Wallace Reis.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Catalyst/View/Component/SubInclude/Visit.pm
package Catalyst::View::Component::SubInclude::Visit;
use Moose;
use Carp qw/croak/;
use MooseX::Types::Moose qw/ Bool /;
use namespace::clean -except => 'meta';

=head1 NAME

Catalyst::View::Component::SubInclude::Visit - visit() plugin for C::V::Component::SubInclude

=head1 VERSION

Version 0.07_03

=cut

our $VERSION = '0.07_03';
$VERSION = eval $VERSION;

=head1 SYNOPSIS

In your view class:

  package MyApp::View::TT;
  use Moose;

  extends 'Catalyst::View::TT';
  with 'Catalyst::View::Component::SubInclude';

  __PACKAGE__->config( subinclude_plugin => 'Visit' );

Then, somewhere in your templates:

  [% subinclude('/my/widget') %]

=head1 DESCRIPTION

C<Catalyst::View::Component::SubInclude::Visit> uses C<< $c->visit() >> to 
render subinclude contents.

This method is only supported when using L<Catalyst> version 5.71000 or newer.

=head1 METHODS

=head2 C<generate_subinclude( $c, $path, @args )>

This is (roughly) equivalent to the following call:

  $c->visit( $path, @args );

But it will handle all the nasty details such as localizing the stash, 
parameters and response body. This is necessary to keep behavior consistent 
with the other plugins.

=head1 CONFIGURATION

=head2 keep_stash

You can choose to not localize the stash for Visits' subinclude calls. The subrequest
will have the same stash as the request that spawned it. Configure the keep_stash key
in your view:

    __PACKAGE__->config(
        subinclude => {
            'Visit' => {
                keep_stash => 1,
            },
        }
    );

Note: changes in the stash during a Visit subinclude will be visible after the include 
returns.

=cut

has keep_stash => (
    isa => Bool,
    is => 'ro',
    default => 0,
);

sub generate_subinclude {
    my ($self, $c, $path, @params) = @_;

    croak "subincludes through visit() require Catalyst version 5.71000 or newer"
        unless $c->can('visit');

    {
        local $c->{stash} = $self->keep_stash ? $c->{stash} : {};
        
        local $c->request->{parameters} = 
            ref $params[-1] eq 'HASH' ? pop @params : {};

        local $c->response->{body};

        my $captures = ref $params[0] eq 'ARRAY' ? shift @params : [];
        $c->visit( $path, $captures, \@params );

        return $c->response->{body};
    }

}

=head1 SEE ALSO

L<Catalyst::View::Component::SubInclude|Catalyst::View::Component::SubInclude>, 
L<Catalyst|Catalyst>

=head1 AUTHOR

Nilson Santos Figueiredo Junior, C<< <nilsonsfj at cpan.org> >>

=head1 SPONSORSHIP

Development sponsored by Ionzero LLC L<http://www.ionzero.com/>.

=head1 COPYRIGHT & LICENSE

Copyright (C) 2009 Nilson Santos Figueiredo Junior.

Copyright (C) 2009 Ionzero LLC.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

__PACKAGE__->meta->make_immutable;
1;


lib/Catalyst/View/Component/SubInclude/SubRequest.pm
package Catalyst::View::Component::SubInclude::SubRequest;
use Moose;
use Carp qw/croak/;
use MooseX::Types::Moose qw/ Bool /;
use namespace::clean -except => 'meta';

=head1 NAME

Catalyst::View::Component::SubInclude::SubRequest - Sub-requests plugin for C::V::Component::SubInclude

=head1 VERSION

Version 0.07_03

=cut

our $VERSION = '0.07_03';
$VERSION = eval $VERSION;

=head1 SYNOPSIS

In your application class:

  package MyApp;

  use Catalyst qw/
    ConfigLoader
    Static::Simple
    ...
    SubRequest
  /;

In your view class:

  package MyApp::View::TT;
  use Moose;

  extends 'Catalyst::View::TT';
  with 'Catalyst::View::Component::SubInclude';

  __PACKAGE__->config( subinclude_plugin => 'SubRequest' );

Then, somewhere in your templates:

  [% subinclude('/my/widget') %]

=head1 DESCRIPTION

C<Catalyst::View::Component::SubInclude::SubRequest> uses Catalyst sub-requests
to render the subinclude contents. 

It requires L<Catalyst::Plugin::SubRequest>.

=head1 METHODS

=head2 C<generate_subinclude( $c, $path, @args )>

This will make a sub-request call to the action specified by C<$path>. Note that
C<$path> should be the private action path - translation to the public path is
handled internally.

So, after path translation, the call will be (roughly) equivalent to:

  $c->sub_request( $translated_path, {}, @args );

Notice that the stash will be empty by default. This behavior is configurable
(see below).

=head1 CONFIGURATION

=head2 keep_stash

You can choose to not localize the stash for SubRequests' subinclude calls. The subrequest
will have the same stash as the request that spawned it. Configure the keep_stash key
in your view:

    __PACKAGE__->config(
        subinclude => {
            'SubRequest' => {
                keep_stash => 1,
            },
        }
    );

Note: the stash that the subrequest recieves is a shallow copy of the original stash. That
means that changes to values of keys on the first level of the stash will be lost when the
subrequest call returns. Don't count on this behaviour, as it may change in the future.

=cut

has keep_stash => (
    isa => Bool,
    is => 'ro',
    default => 0,
);

sub generate_subinclude {
    my ($self, $c, $path, @params) = @_;
    my $stash = $self->keep_stash ? $c->stash : {};

    croak "subincludes through subrequests require Catalyst::Plugin::SubRequest"
        unless $c->can('sub_request');

    my $query = ref $params[-1] eq 'HASH' ? pop @params : {};
    
    my $action = blessed($path)
          ? $path
          : $c->dispatcher->get_action_by_path($path);

    my $uri = $c->uri_for( $action, @params );

    $c->sub_request( $uri->path, $stash, $query );
}

=head1 SEE ALSO

L<Catalyst::View::Component::SubInclude|Catalyst::View::Component::SubInclude>, 
L<Catalyst::Plugin::SubRequest|Catalyst::Plugin::SubRequest>

=head1 AUTHOR

Nilson Santos Figueiredo Junior, C<< <nilsonsfj at cpan.org> >>

=head1 SPONSORSHIP

Development sponsored by Ionzero LLC L<http://www.ionzero.com/>.

=head1 COPYRIGHT & LICENSE

Copyright (C) 2009 Nilson Santos Figueiredo Junior.

Copyright (C) 2009 Ionzero LLC.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

__PACKAGE__->meta->make_immutable;
1;


lib/Catalyst/View/Component/SubInclude.pm
package Catalyst::View::Component::SubInclude;
use Moose::Role;

use Carp qw/croak/;
use Catalyst::Utils ();
use Class::MOP ();
use MooseX::Types::Moose qw/Str HashRef/;
use namespace::clean -except => 'meta';

with 'Catalyst::Component::ContextClosure';

=head1 NAME

Catalyst::View::Component::SubInclude - Use subincludes in your Catalyst views

=head1 VERSION

Version 0.11

=cut

our $VERSION = '0.11';
$VERSION = eval $VERSION;

=head1 SYNOPSIS

  package MyApp::View::TT;
  use Moose;

  extends 'Catalyst::View::TT';
  with 'Catalyst::View::Component::SubInclude';

  __PACKAGE__->config( subinclude_plugin => 'SubRequest' );

Then, somewhere in your templates:

  [% subinclude('/my/widget') %]
  [% subinclude_using('SubRequest', '/page/footer') %]

=head1 DESCRIPTION

C<Catalyst::View::Component::SubInclude> allows you to include content in your
templates (or, more generally, somewhere in your view's C<render> processing)
which comes from another action in your application. It's implemented as a
L<Moose::Role|Moose::Role>, so using L<Moose|Moose> in your view is required.

Simply put, it's a way to include the output of a Catalyst sub-request somewhere
in your page.

It's built in an extensible way so that you're free to use sub-requests,
Varnish ESI (L<http://www.catalystframework.org/calendar/2008/17>) or any other
sub-include plugin you might want to implement.

=head1 STASH FUNCTIONS

This component does its magic by exporting a C<subinclude> coderef entry to the
stash. This way, it's easily accessible by the templates (which is the most
common use-case).

=head2 C<subinclude( $path, @args )>

This will render and return the body of the included resource (as specified by
C<$path>) using the default subinclude plugin.

=head2 C<subinclude_using( $plugin, $path, @args )>

This will render and return the body of the included resource (as specified by
C<$path>) using the specified subinclude plugin.

The C<subinclude> function above is implemented basically as a shortcut which
calls this function using the default plugin as the first parameter.

=head1 SUBINCLUDE PLUGINS

The module comes with two subinclude plugins:
L<SubRequest|Catalyst::Plugin::View::Component::SubRequest>,
L<Visit|Catalyst::Plugin::View::Component::Visit> and
L<ESI|Catalyst::Plugin::View::Component::ESI>.

By default, the C<SubRequest> plugin will be used. This can be changed in the
view's configuration options (either in the config file or in the view module
itself).

    __PACKAGE__->config(
        subinclude_plugin => 'ESI',
        subinclude => {
            'SubRequest' => {
                keep_stash => 1,
            },
            'HTTP::POST' => {
                class => 'HTTP',
                http_method => 'POST',
                ua_timeout => '10',
                uri_map => {
                    '/foo/' => 'http://www.foo.com/',
                },
            },
        },
    );

You can change each plugins' configuration through the keys in the 'subinclude'
config key (example above)

=head2 C<set_subinclude_plugin( $plugin )>

This method changes the current active subinclude plugin in runtime. It expects
the plugin suffix (e.g. C<ESI> or C<SubRequest>) or a fully-qualified class
name in the C<Catalyst::View::Component::SubInclude> namespace.

=head2 Writing plugins

If writing your own plugin, keep in kind plugins are required to implement a
class method C<generate_subinclude> with the following signature:

  sub generate_subinclude {
      my ($class, $c, @args) = @_;
  }

The default plugin is stored in the C<subinclude_plugin> which can be changed
in runtime. It expects a fully qualified class name.

=cut

has 'subinclude_plugin' => (
    is => 'rw',
    isa => Str,
);

has subinclude => (
    is => 'ro',
    isa => HashRef,
    default => sub { {} },
);

around 'new' => sub {
    my $next = shift;
    my $class = shift;

    my $self = $class->$next( @_ );

    my $subinclude_plugin = $self->config->{subinclude_plugin} || 'SubRequest';
    $self->set_subinclude_plugin( $subinclude_plugin );

    $self;
};

before 'render' => sub {
    my ($self, $c, @args) = @_;

    $c->stash->{subinclude}       = $self->make_context_closure(sub { $self->_subinclude( @_ ) }, $c);
    $c->stash->{subinclude_using} = $self->make_context_closure(sub { $self->_subinclude_using( @_ ) }, $c);
};

sub set_subinclude_plugin {
    my ($self, $plugin) = @_;

    my $subinclude_class = blessed $self->_subinclude_plugin_class_instance( $plugin );
    $self->subinclude_plugin( $subinclude_class );
}

sub _subinclude {
    my ($self, $c, @args) = @_;
    $self->_subinclude_using( $c, $self->subinclude_plugin, @args );
}

sub _subinclude_using {
    my ($self, $c, $plugin, @args) = @_;
    $plugin = $self->_subinclude_plugin_class_instance($plugin);
    $plugin->generate_subinclude( $c, @args );
}

has _subinclude_plugin_class_instance_cache => (
    isa => HashRef,
    is => 'ro',
    default => sub { {} },
);

sub _subinclude_plugin_class_instance {
    my ($self, $plugin) = @_;

    my $cache = $self->_subinclude_plugin_class_instance_cache;
    return $cache->{$plugin} if exists $cache->{$plugin};

    my $plugin_config = Catalyst::Utils::merge_hashes(
        $self->subinclude->{ALL}||{},
        $self->subinclude->{$plugin}||{}
    );
    my $short_class = $plugin_config->{'class'} ?
        delete $plugin_config->{'class'}
        : $plugin;
    my $class = $short_class =~ /::/ ?
        $short_class
        : __PACKAGE__ . '::' . $short_class;

    Class::MOP::load_class($class);

    return $cache->{$class} = $class->new($plugin_config);
}

=head1 SEE ALSO

L<Catalyst::Plugin::SubRequest|Catalyst::Plugin::SubRequest>,
L<Moose::Role|Moose::Role>, L<Moose|Moose>,
L<http://www.catalystframework.org/calendar/2008/17>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-catalyst-view-component-subinclude at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-View-Component-SubInclude>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 AUTHOR

Nilson Santos Figueiredo Junior, C<< <nilsonsfj at cpan.org> >>

=head1 CONTRIBUTORS

Tomas Doran (t0m) C<< <bobtfish@bobtfish.net >>.

Vladimir Timofeev, C<< <vovkasm at gmail.com> >>.

Wallace Reis (wreis) C<< <wreis@cpan.org> >>.

Jose Luis Martinez (JLMARTIN) C<< <jlmartinez@capside.com> >>.

=head1 SPONSORSHIP

Development sponsored by Ionzero LLC L<http://www.ionzero.com/>.

=head1 COPYRIGHT & LICENSE

Copyright (C) 2010 Nilson Santos Figueiredo Junior and the above contributors.

Copyright (C) 2009 Nilson Santos Figueiredo Junior.

Copyright (C) 2009 Ionzero LLC.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;
